packages/kbn-zod-helpers/src/required_optional.ts

```diff
@@ -28,7 +28,13 @@
  */
 export type RequiredOptional<T> = { [K in keyof T]-?: [T[K]] } extends infer U
   ? U extends Record<keyof U, [unknown]>
-    ? { [K in keyof U]: U[K][0] }
+    ? {
+        [K in keyof U]: Record<string, unknown> extends U[K][0]
```

Comment:
I think this:
```
export type RequiredOptional<T> = { [K in keyof T]-?: [T[K]] } extends infer U
  ? U extends Record<keyof U, [unknown]>
    ? {
        [K in keyof U]: U[K][0] extends Record<string, unknown> | undefined
          ? undefined extends U[K][0]
            ? RequiredOptional<U[K][0]> | undefined
            : RequiredOptional<U[K][0]>
          : U[K][0];
      }
    : never
  : never;
```
works better, but I'm still not 100% sure that it handles all the type edge cases correctly. Mostly putting it here for posterity in case I or someone else comes back to try this again later.